docs/coding-guidelines.md

```diff
@@ -299,3 +300,55 @@ stack potentially unboundedly. Instead use iteration e.g. via `loop` or `for`.
 - https://en.wikipedia.org/wiki/Tail_call
 - https://stackoverflow.com/questions/65948553/why-is-recursion-not-suggested-in-rust
 - https://stackoverflow.com/questions/59257543/when-is-tail-recursion-guaranteed-in-rust
+
+## Allow Correlating Asynchronous Responses to Their Requests
+
+In an asynchronous context, it is important to enable users to determine the correlation between a
+response and a previous request. For example, if a user requests two new connections to the same
+peer, they should be able to match each new connection to the corresponding previous connection
+request without having to guess.
+
+When providing a **method** where the response to that method is delivered asynchronously through an
+event, either synchronously return a request ID which is later on contained in the asynchronous
+response event, or synchronously return a `Future` that eventually resolves into the response.
+
+``` rust
+fn my_method() -> Id {
+  // ...
+}
+```
+
+``` rust
+fn my_method() -> impl Future<Output = Response> {
+  // ...
+}
```

Comment:
Are we doing this anywhere in rust-libp2p, apart from `Transport::dial`?

Reply:
`Tranport::dial` is the only one that comes to my mind. That said, I would like to experiment with it on the API surface of `NetworkBehaviour` implementations. E.g. a `NotBitswap` `NetworkBehaviour` implementation providing a `request_cid` method that returns a `Future`.

I don't have a strong preference, nor a rule for either of the two (return ID or return Future). Thus I suggest keeping both options open. Objections?